ENH: Add fast C-implementations of Patefield and Boyett algorithm for scipy.stats.random_table.rvs

HDembinski opened the pull request:
#### Reference issue
Planned follow-up on #17237.

#### What does this implement/fix?
This adds implementations of the Patefield and Boyett algorithms in C and the automatic switch to chose the fastest algorithm for a given problem unless the user explicitly requests one of the two.

#### Additional information
The C implementation of Boyett's algorithm replaces the Python version from #17237. It is much faster and the Python version has a bug. A new unit test was added to reveal the bug. Temporarily, the old version is still accessible as method="boyett2" to simplify this review, but the method will be removed when this PR is merged.

Benchmarks were added to compare the performance of rvs methods "boyett", "patefield" and "boyett2", see benchmarks/stats.py.

HDembinski commented:
Plots of the benchmark results.

![image](https://user-images.githubusercontent.com/2631586/200170895-67c03240-f6ad-427c-ad6b-f6898017fc92.png)

- "boyett2" (which will be removed) is almost an order of magnitude slower than "boyett"
- "boyett" runtime scales linearly with total number of entries, and is constant wrt to number of cells
- "patefield" scales roughly logarithmically with total number of entries and linearly with number of cells

HDembinski commented:
I see this error a lot in CI coming from the new C extension that I addded: [`E   ImportError: dynamic module does not define module export function (PyInit_rcont)`](https://github.com/scipy/scipy/actions/runs/3404425969/jobs/5661646761#step:6:425)
I don't see this when I build locally with `python dev.py`. If you could lend me some help on how to reproduce this locally, I could try to find a fix.

ev-br commented:
While your branch builds and passes tests locally for me (ubuntu linux, meson build, python 3.8), the CI error you linked is for the `setup.py` build with a debug build of python:

     $ python3.8-dbg -u runtests.py -n -g -j2 -m fast <snip>

`runtests.py` is the setuptools / distiltils analog of `dev.py`, so for a regular python the relevant incantation is something like `$ python runtests.py -s stats` or `$ python runtests.py -t scipy.stats.tests.test_<testfile w/o extension>`.

It won't be my decision on whether it's already OK to drop setup.py based builds, but I do think we aim for CI to pass.

HDembinski commented:
@ev-br Thank you for the nice explanation and the incantation, I can now reproduce this locally and will fix it.

HDembinski commented:
Ping to @mdhaber @WarrenWeckesser @rkern that this PR is to finish up https://github.com/scipy/scipy/pull/17237

Tests are looking good.

tirthasheshpatel commented:
> **Question 2**: I saw in other submodules that there is usually a branch to handle both the old random number generation API of numpy and the new API. My C code requires the new API. Do I have to support the old API as well?

I think we are moving to NumPy >= 1.19.5 in SciPy 1.10.0 so we are guaranteed to have `numpy.random` header files. So we don't need a branch for old NumPy now! The old code needs to be updated (I have started working on it and will submit a PR soon).

HDembinski commented:
Excellent, I am glad to hear this. Then this patch is ready to go.

tirthasheshpatel commented on scipy/stats/_rcont/_rcont.c:
```diff
@@ -0,0 +1,242 @@
+
+// sanity checks for inputs to rcont1 and rcont2,
+// also sets total number of entries
+int rcont_check(double *n, const double *m, int nr, const double *r, int nc, const double *c)
```
This function is unused. Also, the parameter `m` isn't used anywhere in the function (other than just checking if it is a non-null pointer). Did you forget to call this in the `rcont1` and `rcont2` functions? I think the Python function `_process_parameters` does all these tests. Not sure if this is needed at all.

HDembinski replied:
The function was removed.

tirthasheshpatel commented on scipy/stats/_rcont/_rcont.c:
```diff
@@ -0,0 +1,242 @@
+  This function uses a work space of size N which must be preallocated and
+  initialized with rcont1_init.
+*/
+void rcont1(double *matrix, int nr, const double *r, int nc, const double *c,
```
I can confirm that this exactly follows the FORTRAN code in Boyett's paper.

HDembinski replied:
Good to know, although I had not read Boyett's code when I wrote this. I only discovered in hindsight that this is Boyett's algorithm.

tirthasheshpatel reviewed (commented):
Thanks for working on this @HDembinski! It looks really good as-is.

I just left a few comments on `rcont2`'s implementation so it follows the original paper more closely (I think it should be equivalent to the code you have written). Since this is a lot of code, I have a few suggestions below.

- I think we should test against the original implementation (if it's not too much work). We can probably use the same seed and RNG for that but I feel like it would take a lot of effort to get that working. Alternatively, we can try to cover all the C code so we can ensure all the parts are working and give a correct output. Check out the coverage here: https://app.codecov.io/gh/scipy/scipy/pull/17357.
- Need to handle edge cases and add tests for them. Empty outputs give errors while having only one `[0]` in both row and col sums return garbage output. There might be other edge cases to consider.
- Another opinion on this would be super helpful. @WarrenWeckesser  @mdhaber @chrisb83 can any of you take a look?

tirthasheshpatel commented on scipy/stats/_rcont/_rcont.c:
```diff
@@ -0,0 +1,242 @@
+      }
+      double z = random_standard_uniform(rstate);
+      double nlm;
+    l131:
```
I think we should only focus on implementation and testing in this PR. Please ignore this right now. We can come back to this in a follow-up.

HDembinski commented:
> I just left a few comments on `rcont2`'s implementation so it follows the original paper more closely (I think it should be equivalent to the code you have written). Since this is a lot of code, I have a few suggestions below.
> 
>     * I think we should test against the original implementation (if it's not too much work). We can probably use the same seed and RNG for that but I feel like it would take a lot of effort to get that working. Alternatively, we can try to cover all the C code so we can ensure all the parts are working and give a correct output. Check out the coverage here: https://app.codecov.io/gh/scipy/scipy/pull/17357.

Adding a Fortran implementation that uses the numpy random number generator sounds like a lot of work and I am not fluent in Fortran. I would favor the other option, to use an exhaustive test suite that covers all the edge cases.
 
>     * Need to handle edge cases and add tests for them. Empty outputs give errors while having only one `[0]` in both row and col sums return garbage output. There might be other edge cases to consider.

Could you perhaps commit the test code for these edge cases? I am happy to work on a fix then.

HDembinski commented on scipy/stats/_rcont/_rcont.c:
```diff
@@ -0,0 +1,242 @@
+      {
+        for (int j = m; j < nc - 1; ++j)
+          *ptr(matrix, nr, nc, l, j) = 0;
+        ia = 0;
```
This cannot be removed. After line 177, we break out of the inner loop and arrive at line 204, where ia is assigned to the matrix.

tirthasheshpatel replied:
Oh, OK, I see what's happening. This code wasn't covered which is why I too was unsure whether it would work. It still shows that this part is uncovered. Could you come up with an input that covers this case and add it as a test?

HDembinski replied:
I am trying. It is not obvious.

HDembinski replied:
I tried, but I was not successful so far in constructing test cases that cover 100 % of the C code. We can still do this later.

tirthasheshpatel replied:
Thanks for trying! Yes, we can add more tests later. I don't think this is necessary to do right now.

HDembinski commented on scipy/stats/_rcont/_rcont.c:
```diff
@@ -0,0 +1,242 @@
+      const double ii = ib - id; // forth term
+      if (ie == 0)
+      {
+        for (int j = m; j < nc - 1; ++j)
```
The original code is correct, because after the break in line 178, we jump to line 237 where the last value is set.

tirthasheshpatel replied:
I was just trying to be closer to the original FORTRAN code. But it's fine not to add this.

HDembinski commented on scipy/stats/_rcont/_rcont.c:
```diff
@@ -0,0 +1,242 @@
+{
+  // jwork is folded into matrix using last row
+  double *jwork = ptr(matrix, nr, nc, nr - 1, 0);
+  for (int i = 0; i < nc; ++i)
```
I have to check, but I think you are right, the last entry of jwork is never used until it is overridden at the end of the algorithm in line 241.

HDembinski replied:
I applied the change and added a comment why the loop can skip the last element.

tirthasheshpatel replied:
Looks good, thanks!

mdhaber reviewed (commented):
The new edge case tests look good.

I ran the benchmakrs for `boyett` and `patefield` and I see:
```
[100.00%] ··· =========== ======= ============= ============= ============= =============
              --                                           ncell
              ------------------- -------------------------------------------------------
                 method     ntot        4             64           256           1024
              =========== ======= ============= ============= ============= =============
                 boyett      10      101±6μs       117±2μs       494±7μs     1.37±0.06ms
                 boyett     100      870±20μs      812±5μs     1.23±0.02ms   2.07±0.05ms
                 boyett     1000   7.96±0.09ms    7.29±0.1ms    7.85±0.1ms    8.85±0.1ms
                 boyett    10000    82.0±0.6ms    76.1±0.1ms    76.3±0.6ms    77.6±0.2ms
               patefield     10      472±8μs     6.04±0.02ms    16.4±0.2ms    30.0±0.5ms
               patefield    100      842±10μs     29.0±0.3ms    124±0.9ms      473±4ms
               patefield    1000     419±2μs      27.6±0.2ms    147±0.7ms      538±5ms
               patefield   10000     486±7μs      20.4±0.1ms    118±0.6ms      450±5ms
              =========== ======= ============= ============= ============= =============
```
Is it expected for `boyett` to be faster in all these cases?

HDembinski commented:
> Is it expected for `boyett` to be faster in all these cases?

@mdhaber Yes, that seems to agree qualitatively with the plot I made further up from the results on my machine. Boyett is not always faster, it is only faster than Patefield when the total number of entries in the table is small compared to the number of cells. This is what the results show.

HDembinski commented on scipy/stats/_multivariate.py:
```diff
@@ -5413,34 +5415,34 @@ def _process_rvs_method(cls, method, r, c, n):
-        #     return cls._rvs_patefield
-        # return cls._rvs_boyett
-        # 'fac' has to be estimated empirically
+        fac = 1.0  # benchmarks show that this value is about 1
+        k = len(r) * len(c)  # number of cells
+        # n + 1 guards against failure if n == 0
+        if n > fac * np.log(n + 1) * k:
+            return cls._rvs_patefield
```
The factor varies between 0.8 and 1.6 in my tests. There is no need to fine-tune it. In the worst case, you get a slightly less-than-optimal runtime.

HDembinski commented on scipy/stats/_multivariate.py:
```diff
@@ -5358,8 +5358,8 @@ def _process_parameters(row, col):
         """
-        r = np.array(row, dtype=np.int_, copy=True)
-        c = np.array(col, dtype=np.int_, copy=True)
+        r = np.array(row, dtype=np.double, copy=True)
+        c = np.array(col, dtype=np.double, copy=True)
```
> If the user provides marginals that are beyond the range within which float64 can represent all integers exactly, will our output not necessarily be exact? For instance, suppose a row marginal is 1.8446744073709552e19. Will the row of the tables we return sum exactly to 18446744073709552000, or is it only required that they sum to 1.8446744073709552e19 in float64 precision? (I doubt it matters in most applications, but a note in the documentation might be warranted.)

If you go into the realm where round-off error happens, Patefield's algorithm should still work. I should add a test for this case. The results may not add up exactly the same number of entries then, but it is a tiny relative error that will be negligible in calculations that use those tables.

HDembinski replied:
To clarify, this bit in Patefield would have to convert a lot of numbers from integer to double back and forth
```
      double x = exp(
          lgamma(ia + 1) + lgamma(ib + 1) + lgamma(ic + 1) + lgamma(id + 1) - lgamma(ie + 1) - lgamma(nlm + 1) - lgamma(id - nlm + 1) - lgamma(ia - nlm + 1) - lgamma(ii + nlm + 1));
```
The ia, ib, etc. are elements from the output matrix, so they would be int64 if we make the matrix dtype int64. So we can safe 9 conversions from int to double here.

However, I am also open to replacing the loggamma function with a fast log-factorial function that I discovered in numpy https://github.com/numpy/numpy/blob/main/numpy/random/src/distributions/logfactorial.c
With that we could do the whole calculation efficiently in int64 without converting ints to double.

mdhaber replied:
Replacing and using int throughout sounds good to me if it's OK with Tirth.

rkern replied:
In the C code, please use `long` (or maybe `int64_t`).

tirthasheshpatel replied:
+1 for using `int64_t` and `logfactorial`.

HDembinski replied:
Ok, I will change everything to using int64.

HDembinski commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -2523,11 +2526,11 @@ def test_process_parameters(self):
         message = "each element of `row` must be an integer"
         with pytest.raises(ValueError, match=message):
-            random_table([2.1, 2.1], [1, 1, 2])
+            random_table([2.1, 1.9], [1, 1, 2])
```
This change was necessary to pass the previous check, which checks that the column sum and row sum is equal. Before there were, because the doubles were rounded down automatically in the conversion from double to int. Now that I changed the dtype, the sum is calculated exactly and I had to adjust the numbers.

mdhaber replied:
I see, and you want to make sure it gets to the "must be an integer" error.

HDembinski commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -2651,72 +2636,142 @@ def method(x):
+        rvs = ct.rvs(row, col, size=10000, method=method,
+                     random_state=self.get_rng())
+
+        unique_rvs, counts = np.unique(rvs, axis=0, return_counts=True)
```
Yes. It is a bit of a circular argument, we check pmf with the help of random samples and vice versa. If you have a better idea, let me know. I think it is ok, because it is very unlikely that we make two mistakes which somehow compensate so that pmf and random sampling agree although both are wrong.

mdhaber replied:
I think this is fine, but I think we did something very similar in `test_pmf_logpmf`. If we are doing check the relationship in the opposite direction somehow, that's fine, but maybe I missed it. As an example, if this were a univariate distribution, we might do check `cdf(ppf(p)) = p`, but we should also check `ppf(cdf(x)) = x`. To me, this looked like were applying the same approach twice, though.

HDembinski replied:
The test is very similar, but it is not redundant. I added some comments in the code to explain this. We cannot do the equivalent of `cdf(ppf(p)) = p` and `ppf(cdf(x)) = x` here, because random sample generation and pmf computation are not inverses of each other.

HDembinski commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -2651,72 +2636,142 @@ def method(x):
+        assert_equal(rv, expected)
+
+    @pytest.mark.parametrize('v', (1, 2))
+    def test_rvs_rcont(self, v):
```
During development, this test was useful to check my implementation of rcont1 and rcont2. Now this is tested implicitly by the higher-level tests of random_table. We can remove this, but I thought it cannot hurt to have.

mdhaber replied:
OK. Just noting that as a comment in the code would be good, then, since it confused at least one reader.

HDembinski replied:
Ok, I added that comment.

HDembinski commented:
Let me summarise what is left to do:
- Remove boyett2: No one objected to that and we can all see that rcont1 is the faster implementation.
- Switch row, col and output matrix to dtype int64, rcont1 and rcont2 will calculate in int64

mdhaber commented:
Just a thought that we need to either merge this, fix the rough edges of the existing implementation (e.g. issues with zero columns/rows), or remove `random_table` from the documentation before 1.10 branches in a few days. I don't think we want `random_table` to appear as-is in 1.10, though.

HDembinski commented:
@mdhaber Sorry, I was not around the last two weeks. I will make the last two changes. I hope it is still in time?

mdhaber commented:
I think we have until Tuesday. Let Tirth and I know when it's ready from your perspective.

HDembinski commented:
@mdhaber I implemented the changes. I had to copy the logfactorial code from numpy, because it is not exposed by the numpy public interface. I added the Numpy license to LICENSES_bundled.txt and to the two copied files. Tests pass locally, I still need to fix the CI tests here.

HDembinski commented:
Some builds fail because `-Werror` is set. The warning is generated by pytime.h, which I do not include directly. So there is nothing I can do about it.

This build fails, because `-Werror` is set:
https://github.com/scipy/scipy/actions/runs/3618705994/jobs/6098882214#step:9:1154

This build succeeds, because `-Werror` is NOT set and the tests succeed as well:
https://github.com/scipy/scipy/actions/runs/3618705994/jobs/6098882524#step:4:1255

tirthasheshpatel commented on scipy/stats/_rcont/rcont.pyx:
```diff
@@ -1,18 +1,19 @@
 from numpy.random cimport bitgen_t
 from cpython.pycapsule cimport PyCapsule_GetPointer, PyCapsule_IsValid
 
+ctypedef np.int64_t tab_t
```
I think this re-typedef is causing errors that you are seeing in the build logs. Moving this under the `cdef extern from "_rcont.h"` and using `from libc.stdint import int64_t` fixes the problem.

![image](https://user-images.githubusercontent.com/43181252/205611725-88141619-4422-459f-a1ce-5194e357db16.png)

tirthasheshpatel replied:
I am seeing the warnings again for a rebuild. Can't make sense of why it's happening for only `_rcont`.

HDembinski replied:
I am not a Cython expert, so I don't know whether this is an issue and/or what the correct way is. In my other projects, I wrap my C/C++ code with pybind11 where I know precisely what is happening. I still used Cython here to follow the general trend in Scipy.

HDembinski replied:
Can't we just silence the warning locally? We do that in Boost as a last resort when the warning is a false positive as it is here.

tirthasheshpatel replied:
> Can't we just silence the warning locally? We do that in Boost as a last resort when the warning is a false positive as it is here. If this were just C/C++, I would know how to do this, but here we have Cython in the mix and I don't know how to handle that.

I don't think it would be a good idea to silence the warning even though it's a false positive. Since it's just the Python header messing with the standard lib, we can move the include of `math.h` below other includes and that should do it. Look at this comment for details: https://github.com/scipy/scipy/pull/17357#discussion_r1040276694.

tirthasheshpatel reviewed (commented):
Thanks for the update @HDembinski! The C code looks really good! I made one suggestion to resolve the errors we are seeing in the CI log. This looks really close to completion from my side; I will take a final look once you commit the suggestion.

HDembinski commented on scipy/stats/_multivariate.py:
```diff
@@ -5358,8 +5358,8 @@ def _process_parameters(row, col):
         """
-        r = np.array(row, dtype=np.double, copy=True)
-        c = np.array(col, dtype=np.double, copy=True)
+        r = np.array(row, dtype=np.int64, copy=True)
+        c = np.array(col, dtype=np.int64, copy=True)
```
We do this because the internal C code expects int64 values and row-major order (for C). We enforce this here. I suppose that `copy=True` is not strictly necessary, since `row` and `col` are never mutated, but it can't hurt to have that extra safety and it follows the example of other distributions.

tirthasheshpatel replied:
Sounds good, thanks for the clarification.

HDembinski commented on scipy/stats/_rcont/_rcont.c:
```diff
@@ -0,0 +1,215 @@
+  Author: Hans Dembinski
+*/
+
+#include <math.h>
```
I cannot remove this, because I use `exp` and `floor` from `math.h`.

tirthasheshpatel replied:
Oh, my bad. I should have been more specific. I meant putting the math include after other includes or moving it to the header file. This is the reason:

![image](https://user-images.githubusercontent.com/43181252/205779685-f0e09ec9-504c-4a20-a695-b0d01f8ab0cc.png)

HDembinski commented:
Btw, using logfactorial from numpy made the Patefield implementation about 30 % faster. :)
![image](https://user-images.githubusercontent.com/2631586/205688021-7bed417e-564b-4bc9-babd-bf6d75873a8c.png)

mdhaber commented:
So @tirthasheshpatel what's the status here? SciPy 1.10 branches tomorrow, I think - what does this need to make it in?

tirthasheshpatel commented:
> So @tirthasheshpatel what's the status here? Are we all set except for the build problem?

I think yes. The C code and the implementation look good to me. If the tests look good to you (can you please approve if they do?), I think we can approve and merge once the build error is gone.

mdhaber commented:
@tirthasheshpatel If you see this before @HDembinski is able to make the changes, would you try making them and pushing to see if that works? Then go ahead and merge if you get it to build successfully.

mdhaber commented on scipy/stats/_rcont/_rcont.c:
```diff
@@ -0,0 +1,215 @@
+#include <math.h>
+
+#include "_rcont.h"
+#include "logfactorial.h"
```
```suggestion
#include "logfactorial.h"
#include <math.h>
```

tirthasheshpatel replied:
The change looks good, thanks! One unrelated failure though :(

tirthasheshpatel reviewed (approved):
CI looks (mostly) happy. Two unrelated failures. Otherwise, the C code and the Python implementation LGTM! Thanks for the changes @HDembinski! This looks like a really cool addition to SciPy. And thanks for the reviews @mdhaber @rkern! I can merge once the CI finishes.

HDembinski commented:
Thank you, @mdhaber and @tirthasheshpatel, for your quick action to get this merged into the next release!
